### DISCUSSION TITLE
The database does not match the target database warning after migrations

### DISCUSSION THREAD

**Discussion by viperciyes** [serverpod/serverpod#2773] [Q&A] [answered]
To everyone behind Serverpod: 
`A world of thanks for all the effort to build and maintain this, you are helping many people reach their goals.
`
And here is what I want to ask:
I add a new column to an existing table, I create a migration, and get the warning: 
```
Migration Warnings:
 • One or more columns are added to table "users" which cannot be added in a table migration. The
complete table will be deleted and recreated.
Migration aborted. Use --force to ignore warnings.
```

Then I force it, and run apply migrations, and get this result:

```
Applied database migration:
 - 20240915200556757
WARNING: The database does not match the target database:
 - Table "users" is not like the target database.
 -Hint: Did you forget to apply the migrations (--apply-migrations) or run a repair migration (--apply-repair-migration)?
```

Then I create a repair migration and apply it and the result:

```
SERVERPOD version: 2.1.1, dart: 3.3.4 (stable) (Tue Apr 16 19:56:12 2024 +0000) on "macos_arm64", time: 2024-09-15 20:09:53.552788Z
mode: development, role: maintenance, logging: normal, serverId: default
Database repair migration "20240915200939084" applied.
WARNING: The database does not match the target database:
 - Table "users" is not like the target database.
Hint: Did you forget to apply the migrations (--apply-migrations) or run a repair migration (--apply-repair-migration)?
```
Now, if this is the right way to do these changes?
If I'm doing something wrong, please advise :)
Also, after applying repair migration, I can see the new column in the database despite the warning.

**Comment by Isakdl** [comment, marked as answer]
Question, did you regenerate the code as well?

The integrity check is done by generating the expected table definition in dart and then checking against the real state in the db. If you only created migrations but did not regenerate the code after the model change then you may get a warning like this.

**Comment by viperciyes** [comment]
Right, this time I made the change, run serverpod generate, then created migration --forced and applied migrations. and I didn't get that last warning. are there any guides for these migrations, for cases when we don't want to drop and recreate tables, like when we add a new column, make the existing rows have null values for the new fields, etc. or what is the best/ideal way to manage these changes?

**Reply by SandPod** [reply]
> are there any guides for these migrations, for cases when we don't want to drop and recreate tables, 

We don't have specific guides for this in our documentation at the moment, but following standard practices when working with a Postgres Database works.

Here are two ways to achieve your desired change without dropping the table:
1. Create a new column with a default value. This will automatically be populated for all rows in your database.
2. If utilizing default values is not possible, you could create a new column that is nullable, populate it manually, and then drop the nullability of the column.

Both of these methods should allow you to add a new column without dropping all existing data.
